api/hypershift/v1alpha1/nodepool_types.go

```diff
@@ -888,16 +888,55 @@ type AgentNodePoolPlatform struct {
 	AgentLabelSelector *metav1.LabelSelector `json:"agentLabelSelector,omitempty"`
 }
 
-type AzureNodePoolPlatform struct {
-	VMSize string `json:"vmsize"`
-	// ImageID is the id of the image to boot from. If unset, the default image at the location below will be used:
-	// subscription/$subscriptionID/resourceGroups/$resourceGroupName/providers/Microsoft.Compute/images/rhcos.x86_64.vhd
+type AzureVMImageType string
+
+const (
+	ImageID          AzureVMImageType = "ImageID"
+	AzureMarketplace AzureVMImageType = "AzureMarketplace"
+)
+
+// AzureVMImage represents the different types of image data that can be provided for an Azure VM.
+// +union
+type AzureVMImage struct {
+	// Type is the type of image data that will be provided to the Azure VM. This can be either "ImageID" or
+	// "MarketplaceImageInfo".
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=ImageID;AzureMarketplace
+	// +unionDiscriminator
+	Type AzureVMImageType `json:"azureImageType"`
+
+	// ImageID is the Azure resource ID of a VHD image to use to boot the Azure VMs from.
+	//
 	// +optional
+	// +unionMember, optional
 	ImageID string `json:"imageID,omitempty"`
+
+	// MarketplaceImageInfo contains the Azure Marketplace image info to use to boot the Azure VMs from.
+	//
+	// +optional
+	// +unionMember, optional
+	MarketplaceImageInfo *MarketplaceImage `json:"marketplaceImage,omitempty"`
+}
+
+type AzureNodePoolPlatform struct {
+	// VMSize is the Azure VM instance type to use for the nodes being created in the nodepool.
+	//
+	// +kubebuilder:validation:Required
+	VMSize string `json:"vmsize"`
+
+	// Image is the image to boot the VMs with
+	//
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Image is immutable"
+	// +kubebuilder:validation:Required
```

Comment:
Shouldn't we have a len>0 validation here

Reply:
Since this is a struct I don't think you need a length check